### PR TITLE
LaplaceHypre3d

### PR DESCRIPTION
**Pull request by johnomotani**
Here is an attempt at a 3d Laplace solver using the Hypre interface. @YiningQin @ZedThree @bendudson @acfisher @jonesholger @xxu

It does not pass the integrated test. The first thing I'd like to check is whether this is just due to the solver tolerances - we had to use pretty tight tolerances (rtol=1e-10, atol=1e-13) for the PETSc 3d Laplace solver to pass. I could not find a method to set tolerances in the Hypre interface - @acfisher did I miss something?

I had to change the definition of the `HypreMalloc` to compile with the CPU version, and also change the calls from, e.g. `HypreMalloc(&I, vsize*sizeof(HYPRE_BigInt))` to `HypreMalloc(I, vsize*sizeof(HYPRE_BigInt))` - not sure if the `&` had some significance for the GPU version.

There seems to be a bug in `configure.ac`. I couldn't configure using `--with-hypre` and `--with-petsc` at the same time. When I did, the check for `libHYPRE` failed; it seemed to be because it was trying to link with `-L $(PETSC_LIB)` instead of the path to the Hypre library, but I don't understand why when the check worked without PETSc...

### CODE REVIEW

**Comment by YiningQin** [comment]
Thanks John, that's really helpful. Let's discuss it at Monday's meeting.

Best,

Yining

**Comment by ZedThree** [comment]
There's a unit test on the PETSc 3D AMG that's probably worth copying too

**Comment by johnomotani** [comment]
> There's a unit test on the PETSc 3D AMG that's probably worth copying too

Good point, I've (partially) copied now. Many of the PETSc unit tests use a matrix-vector product, which I couldn't see implemented for Hypre. It would be useful to have though, because the solve-test is failing, saying the calculated solution is zero everywhere. I don't have any idea why, and don't have time to debug this solver any more...

**Comment by johnomotani** [comment]
BTW I had to hack a couple of things on this branch for a non-RAJA build:
on l.24 of `include/bout/hypre_interface.hxx`
```
-#if 1   //HYPRE with Cuda enabled : todo use appropriate flag
+#if 0   //HYPRE with Cuda enabled : todo use appropriate flag
```
on l.16 of `single_index_ops.hxx`
```
-using EXEC_POL = RAJA::loop_exec;
+//using EXEC_POL = RAJA::loop_exec;
```
and somewhere in the unit tests (with these hacks, don't know if it's related) there's a segfault. I worked around that by using `--gtest_filter` to only run the `LaplaceHypre3d` tests.

File: include/bout/hypre_interface.hxx [hunk 1]
```diff
@@ -130,8 +130,8 @@ public:
 
     location = f.getLocation();
     initialised = true;
-    HypreMalloc(&I, vsize*sizeof(HYPRE_BigInt)); 
-    HypreMalloc(&V, vsize*sizeof(HYPRE_Complex));
+    HypreMalloc(I, vsize*sizeof(HYPRE_BigInt));
+    HypreMalloc(V, vsize*sizeof(HYPRE_Complex));
```
Comment (jonesholger) [inline include/bout/hypre_interface.hxx:134]: @johnomotani In cudaMallocManaged mode of the define the call is expecting a void** (i.e. Handle), so some finesse still needed:  

__host__ ​cudaError_t cudaMallocManaged ( void** devPtr, size_t size, unsigned int  flags = cudaMemAttachGlobal )

Comment (johnomotani) [reply include/bout/hypre_interface.hxx:134]: Ah, OK. Thanks @jonesholger. Trouble with me not being able to test the GPU version...  Without these changes the CPU version refused to compile. Please feel free to push fixes to this branch!

**Comment by johnomotani** [comment]
This should possibly be working and passing tests now (but see #2225, which also applies to the tests here).

@ZedThree - this is so similar to LaplacePetsc3dAmg that it would be really nice to remove the code duplication (otherwise maintenance will be a headache!). One way to approach that might be to add a `PetscSystem` similar to Aaron's `HypreSystem` so we can move the PETSc calls `KSPSolve()`, etc.) under there. Then I hope it would be relatively simple to template a 3d Laplace solver on the PETSc and Hypre interfaces, or make the interfaces implementations of an abstract `LinearSolver` so `Laplace3d` doesn't need to know which one it's using..
It would be nice to merge the integrated test too, so there's only one `test-laplace3d` to maintain, which runs both implementations (if they're available).
I'm less sure about merging the unit tests (`test_laplace_hypre3d.cxx` and `test_laplace_petsc3d.cxx`). At the moment they are different because the interface to matrix-vector multiplication is different, but probably that could be massaged away. I don't know how gtest would play with templated tests? Maybe it's easier if the interfaces are implementations of an abstract `LinearSolver`, and the tests can just be parameterised on the type?

**Comment by johnomotani** [comment]
Hacks still needed from the `next-hypre-outerloop-cuda-debug` branch, which could do with a better fix somewhere (but aren't part of the LaplaceHypre3d solver):
* Commenting out `HypreFinalize()` in the `HypreSystem` class to stop unit tests crashing. Probably the correct solution is to create a HypreLib reference counter like the PetscLib one.
* Comment out `using EXEC_POL = RAJA::loop_exec;` in `single_index_ops.hxx` to compile without RAJA

**Comment by johnomotani** [comment]
Closing since @jonesholger merged the changes manually.

**Pull request closed without merging**